src/librustc/hir/def.rs

```diff
@@ -129,6 +129,34 @@ impl DefKind {
             _ => "a",
         }
     }
+
+    pub fn is_valid_in_ns(&self, ns: Namespace) -> bool {
+        match self {
+            DefKind::Mod
+            | DefKind::Struct
+            | DefKind::Union
+            | DefKind::Enum
+            | DefKind::Variant
+            | DefKind::Trait
+            | DefKind::OpaqueTy
+            | DefKind::TyAlias
+            | DefKind::ForeignTy
+            | DefKind::TraitAlias
+            | DefKind::AssocTy
+            | DefKind::AssocOpaqueTy
+            | DefKind::TyParam => Namespace::TypeNS == ns,
```

Comment:
```suggestion
            | DefKind::TyParam => ns == Namespace::TypeNS,
```
Style nit, here and in few other places.